Add `mbedtls_pk_get_key_type()`

davidhorstmann-arm opened the pull request:
Introduce a new function, `mbedtls_pk_get_key_type()`, to get the PSA key type of a PK context.

~Note: Currently this uses `mbedtls_pk_type_t` and `mbedtls_pk_get_type()` internally, but I intend to refactor it to avoid using this soon.~

Fixes #509

## PR checklist

Please remove the segment/s on either side of the | symbol as appropriate, and add any relevant link/s to the end of the line.
If the provided content is part of the present PR remove the # symbol.

- [x] **changelog** provided
- [x] **framework PR** not required
- [x] **mbedtls development PR** not required because: Crypto changes only
- [x] **mbedtls 3.6 PR** not required because: New API so will not backport
- **tests**  provided

davidhorstmann-arm committed:
Add mbedtls_pk_get_key_type() testcase

Add testcase for a new function, mbedtls_pk_get_key_type(), which will
return the PSA key type of a PK context.

This testcase exercises the function on a PK context that has been
created by wrapping a PSA key with mbedtls_pk_wrap_psa().

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Silly implementation of mbedtls_pk_get_key_type()

Just return PSA_KEY_TYPE_RSA_KEY_PAIR always, as this passes the tests.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Implement the getting of key type properly

Get the key type from the priv_id field of the PK context for wrapped
key pairs. This now passes all of the tests.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Add testcase for key type of parsed keys

Try getting the PSA key type of keys that have been parsed from a file.

For now, just test keys of type PSA_KEY_TYPE_RSA_KEY_PAIR

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Add testcases for elliptic curve key files

Add testcases to read EC keypairs and check their types.
Since we have them available in framework/data_files, use one SECP256r1
key and one Brainpool256r1 key.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Add pk_wrap_psa testcases for EC keys

Add testcases to create 2 different kinds of EC keys in PSA and wrap
them in a PK context before getting their type.

This should defeat the silly implementation of
mbedtls_pk_get_key_type().

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Add testcase to get type of a parsed public key

Parse a public key from a file with mbedtls_pk_parse_public_keyfile()
and get its type.

For now, just do this with RSA (partly because we have server1.key
handy).

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
When a key is public-only, infer its key type

When a PK context only represents a public key it is not backed by a PSA
key so we must infer its PSA key type by inspecting the internal state
of the PK object.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Add testcases for parsed public EC keys

Add testcases to parse public keys for 2 different curves (SECP256r1 and
brainpool256r1) and check that the correct PSA key type is returned.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Add testcases for key pairs copied from PSA

Test mbedtls_pk_get_key_type() on pk contexts created via
mbedtls_pk_copy_from_psa() that store a key pair.

The test function is similar to the one for wrapped PSA keys except
that:
* We need to make the PSA keys exportable
* (Obviously) We call copy_from_psa() instead of wrap_psa()

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Add tests for copy_from_psa() of a public key

Add tests that create a public key in PSA and then use pk_copy_from_psa()
to populate a PK context before getting its type.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Test key type from pk_copy_public_from_psa()

Add a test that creates a PK context using
mbedtls_pk_copy_public_from_psa() (i.e. copying the public part of a PSA
key pair into a PK context) and gets its PSA key type.

Test with 2 types of elliptic curve and RSA, as before.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Add testcase for empty PK context

Test calling mbedtls_pk_get_key_type() on an empty initialized context.
This should return PSA_KEY_TYPE_NONE.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Always zero-initialize key ids

This prevents a sensible compiler warning: if we fail to initialize PSA,
we would be calling psa_destroy_key() on uninitialized memory.

Initialize keys to 0, which is the psa_key_id_t equivalent of NULL for
pointers - calling psa_destroy_key(0) is a noop.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Add PK context member for PSA key type

Add a new member to the mbedtls_pk_context struct to store the PSA key
type of the PK object.

For public keys, which are not backed by a real PSA key, this field
must still be populated with the key type that the key would have if
stored in PSA.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Make mbedtls_pk_get_key_type() a simple getter

Replace the implementation of mbedtls_pk_get_type() with a simple getter
function, now that we have a PSA key type field pk->psa_type.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Populate psa_type in mbedtls_pk_wrap_psa()

Populate the new psa_type field when wrapping a PSA key in a PK context.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Set psa_type correctly in copy_from_psa()

In mbedtls_pk_copy[_public]_from_psa(), set the psa_type field of the PK
context based on the key type of the PSA key that is being copied.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Add extra tests for PK parsing codepaths

Add tests for parsing unencrypted PKCS8 keys and keys in DER format.
This provides better coverage of the myriad parsing possibilities in
mbedtls_pk_parse_key().

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Set the PSA key type when parsing a private key

In mbedtls_pk_parse_key(), set the PSA key type based on the internal
PSA private key, once we have finished parsing.

For simplicity, do this just before we return on the success path.
However, since there are a lot of points in mbedtls_pk_parse_key() where
we return on the success path, make a static helper function and call
that at all of those points.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Add DER testcases for public-key parsing

Add testcases that exercise getting the key type of a PK object created
using mbedtls_pk_parse_public_key() on a DER-encoded file (RSA and ECC)
as this exercises a different codepath.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Set the PSA key type for public key PK objects

When parsing a public key, set the PSA key type to be the correct type
for the public key. This is done separately for RSA and ECC keys.

Note that this is the PSA type that the public key would have if it were
stored in PSA (however, it is not).

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Add missing dependency: MBEDTLS_PK_PARSE_C

This is needed by the testcases that call
mbedtls_pk_parse[_public]_key().

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Add ChangeLog entry for mbedtls_pk_get_key_type()

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Add missing call to mbedtls_free() in testcase

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Add missing initialisation of psa_type

The mbedtls_pk_init() function initialises fields directly (rather than
being a blanket call to memset() or similar) so we must add the new
field to it.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm commented:
Rebased and conflicts resolved, this should be ready for review!

davidhorstmann-arm committed:
Add missing test dependencies

Some tests need a dependency on RSA key generation for setup.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Add missing dependencies for ECC key generation

As with RSA, some tests for ECC keys need support for key pair
generation and therefore require a dependency on it.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Fix typo in casting

The argument holding the PSA key type was mistakenly cast to a
psa_key_id_t rather than a psa_key_type_t.

Fix this to do the correct cast to a psa_key_type_t.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Switch from psa_key_id_t to mbedtls_svc_key_id_t

This ensures that we do the right thing when we are running a PSA server
of some kind where svc_key_id_t encodes information about the owner of
the key in question.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Add missing curve-type dependencies

The testcases were missing dependencies for the exact curves used, which
caused test failures whenever these curves were not available. Correct
this by adding the correct dependencies.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

mpg commented:
Just a general comment as this is the first time I'm looking at this PR since the approach changed: the fact that you've had to add calls to `mbedtls_pk_set_psa_type_from_private()` in so many places in `pkparse.c` makes me a bit nervous: how do we know we haven't forgotten one, or that we won't forget it when adding a new code path in the future (which we probably will some time in 2026 for ML-DSA)?

Hopefully the tests should catch this, but that just move the problem to a different place: how do we know we're testing all cases, and that we will keep doing so in the future?

I wonder if we could structure the code differently to mitigate these risks. I'm thinking of two different approaches:
- From the bottom: can we find places in `pk_rsa.c` and `pk_ecc.c` that would be the single point (for each key type) where `priv_id` gets populated while parsing a key, and set `psa_type` there? (That is, the way you did it for public keys.)
- From the top: rename the existing public (or semi-public) parsing function to `_internal` and make the public function a wrapper that just calls the internal function then `mbedtls_pk_set_psa_type_from_private()`.

Wdyt?

davidhorstmann-arm commented:
I'm happy with either of those two approaches, but the approach from the top seems to be the more robust.

If we were to do the first (from the bottom) approach and then add ML-DSA, we'd need to add a new call to `mbedtls_pk_set_psa_type_from_private()` for the new key type, which we might miss.

I think the 'right' thing to do here would be to see if `mbedtls_pk_parse_key()` can be restructured so that it's easier to do post-parsing actions (and to make it easier to follow the control flow, as it's quite difficult to read at the moment). However, this would take too much time and I'm sure there would be some complications to it.

valeriosetti commented:
Actually I would be in favor of an approach "from the bottom". Unless I missed something, in the end `mbedtls_pk_[ecc|rsa]_set_key` is always called for setting the key from `mbedtls_pk_parse_key`. Can't we set the `psa_type` there?
This would be more uniform with what is done in case of public keys.

With this solution you can also remove change in `copy_from_psa` because that one calls `mbedtls_pk_[ecc|rsa]_set_key` as well.

`mbedtls_pk_[ecc|rsa]_set_key` are the only point where they key pair is set, so setting `psa_type` should guarantee that we're not missing any corner case.

> mbedtls_pk_parse_key() can be restructured ... it's quite difficult to read at the moment.

Agree

valeriosetti commented on tests/suites/test_suite_pk.function:
```diff
@@ -2164,3 +2164,255 @@ exit:
+    psa_set_key_usage_flags(&attrs, 0);
+    psa_set_key_algorithm(&attrs, PSA_ALG_NONE);
+
+    PSA_ASSERT(psa_generate_key(&attrs, &key));
```
Generating keys at every iteration might slow down test case. I think that `pk_psa_setup()` can be helpful here (and also in `pk_get_key_type_copy_from_psa_keypair`, `pk_get_key_type_copy_from_psa_public` and `pk_get_key_type_copy_public_from_psa`). It uses predefined keys so you don't need to generate them every time. This also relaxes requirements for the test function as key generation is no more required (i.e. no need for `PSA_WANT_KEY_TYPE_[ECC|RSA]_KEY_PAIR_GENERATE`).

valeriosetti commented on drivers/builtin/src/pkparse.c:
```diff
@@ -927,6 +927,22 @@ MBEDTLS_STATIC_TESTABLE int mbedtls_pk_parse_key_pkcs8_encrypted_der(
+    }
+
+    pk->psa_type = psa_get_key_type(&attributes);
+
```
If we keep this solution (i.e. we don't opt for the "[from bottom"](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/532#issuecomment-3491235988) approach), no `psa_reset_key_attributes` before returning?

mpg replied:
I think last time I asked Gilles, he said calling `reset_key_attributes()` is good practice but it's not an actual bug if we don't call it.

gilles-peskine-arm replied:
I confirm. Since 3.6.0, there are no cases where `psa_reset_key_attributes()` does anything more than `memset(0)`.

davidhorstmann-arm replied:
> If we keep this solution

In the end, we did not, so resolving this.

valeriosetti commented on include/mbedtls/pk.h:
```diff
@@ -403,6 +409,21 @@ int mbedtls_pk_can_do_psa(const mbedtls_pk_context *pk, psa_algorithm_t alg,
+ * \brief           Get the PSA key type corresponding to the key represented
+ *                  by the given PK context.
+ *
+ * \param pk        The context to query.
```
Should we mention that the context must have been at least initialized first?

davidhorstmann-arm committed:
Change RSA key sizes to 1024

This allows the test cases to run when we are testing the static key
slots in the test_crypto_with_static_key_slots component. That component
deliberately sets a static key slot size that prevents key pairs larger
than RSA 1024.

We could try and gate the tests based on whether we are testing static
key slots but it seems better to retain the coverage. The number of bits
is a parameter of the testcase, so it's easy to change if we change our
mind.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Correct documentation on RSA key size calculation

The documentation comment states that the calculated maximum size
accounts for 7 half-size integers and 2 full-size integers, which would
be 7 + 2*2 = 11 half-size integers altogether.

This is a mistake, as it should be 5 half-size integers and 2 full-size
integers, which makes up the total of 5 + 2*2 = 9 half-size integers
total, as the existing documentation states.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Fix incorrect calculation of ASN1 max size

For bit lengths that are not a multiple of 8, the macro
PSA_KEY_EXPORT_ASN1_INTEGER_MAX_SIZE calculated the incorrect result (it
did integer division by 8 rather than an integer ceil() operation).

This would be fine if we were always using multiples of 8, however to
calculate the maximum size of RSA keys we use a non-multiple of 8 to
allow extra room. This should generate an extra byte in the size.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm commented:
In the last 2 commits I have fixed a couple of very small size calculation issues that I noticed when fixing CI failures.

This is a little cheeky of me as they are technically unrelated to this PR, so I will happily remove those commits if requested.

mpg commented on include/psa/crypto_sizes.h:
```diff
@@ -748,9 +748,14 @@
+ * Perform an integer ceil() operation on the bit length to get the required
+ * byte length, in case a non-exact number of bytes is supplied:
+ *
+ * ceil(x/8) = 1 + (x - 1) / 8
```
Nit: I'm more used to ceil(x/8) = (x + 7) / 8. Of course, both are equivalent, unless x is near the end of the type's range: that is, your formula may have a problem for `x == 0` while mine may have a problem when `x > UINT_MAX - 7`. Here we already assume `bits <= 2^19 - 9` (see above) and in general I think we already assume `UINT_MAX >= 2^32 -1` so we should be good. OTOH, `x == 0` seems a bit unlikely but not explicitly ruled out by existing comments. Wdyt?

davidhorstmann-arm replied:
Nullified by other changes (the old formula was correct).

mpg commented on include/psa/crypto_sizes.h:
```diff
@@ -748,9 +748,14 @@
  */
 #define PSA_KEY_EXPORT_ASN1_INTEGER_MAX_SIZE(bits)      \
-    ((bits) / 8u + 5u)
+    (1u + ((bits) - 1u) / 8u + 5u)
```
I think the previous formula was correct, for reasons that are not obvious from the comment:
- if bits is a multiple of 8, then `bits / 8` is correct;
- if bits is not a multiple of 8, then `bits / 8` underestimates by 1, but OTOH we know for sure we won't need a leading 0 (because the top bit of the top byte is unset), so the `+5u` at the end overestimates by 1, which exactly compensates for the underestimation in the division by 8.

If you agree with this reasoning, I think we can restore the previous formula, but update the comment because we really shouldn't leave such things to the reader.

davidhorstmann-arm replied:
Yes, that sounds correct to me, I'll update it.

mpg reviewed (commented):
Just reviewing the last two commits. Will review the rest as my next step.

mpg commented:
> In the last 2 commits I have fixed a couple of very small size calculation issues that I noticed when fixing CI failures.
> 
> This is a little cheeky of me as they are technically unrelated to this PR, so I will happily remove those commits if requested.

For reference, in general when we notice and fix an issue unrelated to the original goal of the PR, we should ask ourselves:

1. Should this be backported? (That is: (a) does it apply to an LTS branch, and (b) is it worth it?)
2. Am I fixing an observable bug? If so, this needs a ChangeLog entry and regression tests (and backports unless it's new in this branch). (Also, consider if this has a security impact.)

This is especially relevant when the original goal of the PR is something that's development-only and/or doesn't require a ChangeLog.

Here, after reviewing the changes, I think the only issues were in the comments, so the answer to (2) would be no, and regarding (1) I'm not sure but I'd say yes (especially the 2nd one which looks really suspicious and had me a bit worried for a moment).

mpg commented:
> I'm happy with either of those two approaches, but the approach from the top seems to be the more robust.
> 
> If we were to do the first (from the bottom) approach and then add ML-DSA, we'd need to add a new call to `mbedtls_pk_set_psa_type_from_private()` for the new key type, which we might miss.

Intuitively, I think I had a preference for the approach "from the bottom", for the same reasons as Valerio. But you have a point that it means when we add ML-DSA we have to remember to set `psa_type` in the equivalent "bottom" function - or equivalently, we have to remember adding a test that calls `pk_get_key_type()` on a parsed ML-DSA key.

> Hopefully the tests should catch this, but that just move the problem to a different place: how do we know we're testing all cases, and that we will keep doing so in the future?

Well, if I may answer my own question: what if we take the existing test functions for parsing and add a mandatory parameter for "expected key type" (or just a check if the function is already specialised for one type)? I mean, presumably (hopefully!) we already have tests covering all cases that can happen during parsing, so we can just take advantage of that. Also, when we add support for parsing ML-DSA keys, surely we'll want to add test cases using `pk_parse_key()` and if this has a parameter `expected_key_type` (that it checks against calling `pk_get_type()` on the parsed key) then we can't forget to provide it.

Wdyt?

mpg commented:
Alternative approach: `psa_type` is only set for public keys, but for private keys `pk_get_type()` always retrieves it from `priv_id`.

Dream: we store all keys (public or private) in `priv_id`, then we don't need a separate `psa_type` field. I'm explicitly _not_ suggesting we make that change in this PR though :) But I think once we've progressed enough on the cleanups planned for this quarter, that's something I'd like to consider (should be easier to do by then).

davidhorstmann-arm committed:
Restore previous calculation but fix comment

The previous calculation of the size taken by an integer was correct.
However, the comment explanation did not quite account for all cases.
Rewrite the explanation to account for the missed case (a
non-multiple-of-8 bit length) and restore the previous behaviour.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm commented:
> Alternative approach: `psa_type` is only set for public keys, but for private keys `pk_get_type()` always retrieves it from `priv_id`.

That would work well, I think! Should I change the name of the new field to something like `pk->public_psa_type` in that case?

mpg commented:
> Should I change the name of the new field to something like `pk->public_psa_type` in that case?

Yes, I think that would be a better name if we go with this approach!

gilles-peskine-arm commented:
> Well, if I may answer my own question: what if we take the existing test functions for parsing and add a mandatory parameter for "expected key type" (or just a check if the function is already specialised for one type)?

I don't think we need even that. I would suggest, rather, that every parsing function should call a test function `pk_exercise_key()`, and that function should make some consistency checks which include validating the `psa_type` field against the content (and complains if it sees a type or a content format that it doesn't know about).

> Dream: we store all keys (public or private) in priv_id, then we don't need a separate psa_type field.

I think we should always have `psa_type`. It's a straightfoward way to know what the context contains. It costs very little memory and it avoids having to make a client-server call each time we want the type when PSA is client-server. Even when PSA is local, it avoids having to allocate an attribute structure for a call to `psa_get_key_attributes`. `psa_type` should be 0 only for a blank context.

mpg commented:
> I would suggest, rather, that every parsing function should call a test function `pk_exercise_key()`, and that function should make some consistency checks which include validating the `psa_type` field against the content (and complains if it sees a type or a content format that it doesn't know about).

 I like that approach. Then actually every test function that's about a key creation method (not just parsing, also wrapping or copying from psa) can call this "consistency checks" function, which also serves as executable documentation for the invariants of a populated `pk_context` structure.

However, this might be a bit more work and I'd like to avoid delaying this PR, as there are other tasks that depend on it. So here's my suggestion:
1. In this PR: always set `psa_type`, using the approach "from the bottom" favoured by [myself](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/532#issuecomment-3490688329) and [Valerio](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/532#issuecomment-3491235988), but don't expand the testing beyond what's already in this PR (ie, we're confident this is correct now, but might not be fully confident this will remain correct when we add new key types).
2. In a follow-up PR, create a "consistency checks" function in `test_suite_pk.function` and use it in all existing test functions that are about parsing, or other methods of key creation (ie, future-proof our testing).

davidhorstmann-arm committed:
Switch approach for setting psa_type

For key pairs, switch approach for setting the PSA key type. Instead of
setting the key type in many arbitrary places, set it in only two places
(when parsing key pairs):
* mbedtls_pk_rsa_set_key(), for RSA key pairs
* mbedtls_pk_ecc_set_key(), for EC key pairs

This limits the complexity and reduces the number of places where the
key type is assigned, so that it's easier to ensure we do not miss one.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Use pk_psa_setup() instead of psa_generate_key()

This increases the performance of the tests as they do not have to
generate a new key every time but can use pre-generated ones.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Remove unnecessary _GENERATE dependencies

Now that we are no longer generating the keys, we do not need a
dependency on the relevant _GENERATE config option.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Add a parameter to mbedtls_pk_ecc_set_pubkey()

This function is sometimes used to build an ordinary public key and
sometimes used to fill in the public part of a key pair, so it cannot
reliably set the PSA key type.

As a (perhaps clumsy) fix to this, add a parameter to specify whether
the function is being used to fill in the public part of an existing key
pair (is_keypair) and use that to decide whether to update the PSA key
type.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Mention initialization of the context

Point out that the PK context passed to mbedtls_pk_get_key_type() must
have been initialized before the function is called.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm commented:
I believe I have addressed everything here, it should be ready for another round of review.

One thing I am unsure of the approach on is this: The function `mbedtls_pk_ecc_set_pubkey()` is used in 2 situations:
1. To set the public key when parsing a plain public key. In this case we want to update `pk->psa_type` to `PSA_KEY_TYPE_ECC_PUBLIC_KEY(family)`
2. To populate the public key part of an existing key pair in an extra post-parsing step. In this case we want to leave `pk->psa_type` as it is, since it is set already by `mbedtls_pk_ecc_set_key()` to `PSA_KEY_TYPE_ECC_KEY_PAIR(family)`

I have solved this by adding a boolean parameter `is_keypair` to `mbedtls_pk_ecc_set_pubkey()` (allowed as it is internal-only) but reviewers may disagree as to whether this is a suitably neat solution.

davidhorstmann-arm committed:
Rectify key ID server oversight

Use MBEDTLS_SVC_KEY_ID_INIT instead of PSA_KEY_ID_NULL as that is the
type accepted by pk_psa_setup() (and it is needed to take account of
key-id-encodes-owner situations).

Signed-off-by: David Horstmann <david.horstmann@arm.com>

mpg reviewed (dismissed):
LGTM, thanks!

valeriosetti commented on include/psa/crypto_sizes.h:
```diff
@@ -784,7 +788,7 @@
  * - 4 bytes of SEQUENCE overhead;
  * - 3 bytes of version;
- * - 7 half-size INTEGERs plus 2 full-size INTEGERs,
+ * - 5 half-size INTEGERs plus 2 full-size INTEGERs,
```
While reading about this fix, I checked also the line below and it says `7 bytes for the public exponent`, but then above it's `publicExponent INTEGER, -- 32-bit`. The two don't match IMO. Wdyt?

davidhorstmann-arm replied:
I believe this is correct because the 32-bit Integer is encoded in ASN.1 format, so we have:
* 1 byte for the tag
* 1 byte for the length
* 5 bytes for the value (allowing for the possibility that it is unsigned 0xFFFFFFFF, which would require a leading zero byte because of signed-ness)

valeriosetti replied:
Oh good point! Thanks!

valeriosetti commented on include/mbedtls/pk.h:
```diff
@@ -134,6 +134,12 @@ typedef struct mbedtls_pk_context {
+    /* The PSA key type of the key represented by the context.
+     *
+     * Note: Valid even if the PK object is not backed by a PSA key
+     * (as is the case for public keys). */
```
Perhaps it's just me not reading it properly, but the way it's written now seem to suggest that there are other possibilities a part from public keys to get a PK context which is not backed by a PSA key, but this is not correct.
I would have rephrased it as:
```suggestion
     * Note: Valid even for public keys which are not backed by a PSA key */
```

This is really minor though and perhaps I overthought about it, so if you think this is not important feel free to ignore it :)

davidhorstmann-arm replied:
I've reworded it with a comma for extra clarity 🙂

valeriosetti commented on drivers/builtin/src/pk_ecc.c:
```diff
@@ -163,6 +165,12 @@ int mbedtls_pk_ecc_set_pubkey(mbedtls_pk_context *pk, const unsigned char *pub,
+    if (!is_keypair) {
+        pk->psa_type = PSA_KEY_TYPE_ECC_PUBLIC_KEY(pk->ec_family);
+    }
+
```
I'm not a huge fan of the addition of the `is_keypair` parameter. It caused changes in drivers (ex: pkparse) and test code and IMO it's not strictly necessary. Similar scenarios are already handled in other functions, for example:
- `mbedtls_pk_ecc_set_group` [here](https://github.com/Mbed-TLS/TF-PSA-Crypto/blob/19f22a994da0f8980f35bb6beec7461ed1fb6b7a/drivers/builtin/src/pk_ecc.c#L26)
- `mbedtls_pk_rsa_set_key` [here](https://github.com/Mbed-TLS/TF-PSA-Crypto/blob/19f22a994da0f8980f35bb6beec7461ed1fb6b7a/drivers/builtin/src/pk_rsa.c#L46)

I know that in this case `mbedtls_pk_ecc_set_key` set the key-pair key type while `mbedtls_pk_ecc_set_pubkey` tries to set the public key type and that the 2 are not the same, but wouldn't this patch [pk_ecc.c.patch](https://github.com/user-attachments/files/23655189/pk_ecc.c.patch) do the same without extra `is_keypair` parameter?

valeriosetti replied:
On a second thought another thing that seems suspect to me is: why doesn't RSA keys need this fix?

mpg replied:
I think it's because in RSA the keys are encoded differently: the private key encoding (n, e, d, p, q, dp, dq, qp) is naturally a superset of the public key encoding (n, e). While with ECC the private key can be encoded just on its own (d) and optionally the public part (Q) can be encoded, or it will be recomputed from the private key (Q = d * G). So for private keys we just parse d, and only then set Q as a 2nd step.

Not sure if that makes sense?

valeriosetti reviewed (changes requested):
A part from really 2 minor suggestions, I'm not super happy about the `is_keypair` parameter in `mbedtls_pk_ecc_set_pubkey` so I proposed an alternative solution (through a patch). Let me know what you think about it?

davidhorstmann-arm committed:
Remove is_keypair parameter

Instead of explicitly specifying whether we are setting the public part
of a keypair, look to see whether the PSA type has been set already. If
it has been set, leave the existing key type in place as long as it
is the keypair type corresponding to the public key we are setting.
Otherwise, return an error.

Co-authored-by: Valerio Setti <vsetti@baylibre.com>
Signed-off-by: David Horstmann <david.horstmann@arm.com>

davidhorstmann-arm committed:
Reword comment about PSA type in public keys

This makes it a little clearer that public keys are the only case where
the key is not backed by a PSA key.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

valeriosetti commented on tests/suites/test_suite_pk.data:
```diff
@@ -1494,3 +1494,98 @@ pk_copy_public_from_psa:"587CF7C57EB7C6254CBF80CC59846521B4FBCBA8BC4B362A9B043F0
+depends_on:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_BASIC:PSA_WANT_ECC_SECP_R1_256
+pk_get_key_type_parse_key:"../framework/data_files/ec_prv.pk8param.der":PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1)
+
+PK get type from parsed key, RSA, DER
+depends_on:PSA_WANT_KEY_TYPE_RSA_KEY_PAIR_BASIC
+pk_get_key_type_parse_key:"../framework/data_files/server1.key.der":PSA_KEY_TYPE_RSA_KEY_PAIR
```
Initially I thought this was a duplicate of the test on line 1510, but then I realized that this uses files in DER format whereas the other is in PEM format. So now the question is instead: don't we need `MBEDTLS_PEM_PARSE_C` dependency on tests using PEM format? And also: why CI is not failing for this?

valeriosetti reviewed (changes requested):
It's almost OK to me. I only have 1 question about a dependency in some tests. Once this is resolved I will approve.

davidhorstmann-arm committed:
Add missing depencies on MBEDTLS_PEM_PARSE_C

The testcases in question parse data files that are in PEM format, so we
need PEM parsing turned on.

Signed-off-by: David Horstmann <david.horstmann@arm.com>

valeriosetti reviewed (approved):
Thanks for your patience in addressing all my comments. LGTM now!

mpg reviewed (approved):
LGTM, thanks!
